Fix crash when running with --list-listeners and no registered listeners

loximann opened the pull request:
## Description
Running with `--listeners` without registered listeners caused Catch2 programs to crash.

```
$ ./a.out --list-listeners
zsh: segmentation fault (core dumped)  ./a.out --list-listeners
```
The reason was due to dereferencing an iterator pointing past the end of a vector when the vector is empty.

Commit:
Fix crash when running with --list-listeners and no registered listeners.

loximann commented:
Should I add a test case to cover listing when no listeners are registered?

horenmar commented:
>  Should I add a test case to cover listing when no listeners are registered?

Yes.

Commit:
Added failing test case.

loximann commented:
I modified the fix so that defaultListListeners is single return (which in general I think makes things cleaner). With that, no test cases are needed for full coverage; however I added a test case to avoid regressions.

horenmar commented:
Please keep the early exit.

And coverage isn't really important, the regression check is.
